### PR TITLE
Config

### PR DESCRIPTION
**Pull request by ashketchupppp**
This pull request adds a number of things for configuring the bot from a config file, as well as a couple other design changes
Config changes:
- `main.go` has a default configuration struct which is writes to JSON if it can't find a config file
- Added some single interfaces for , relevant to configuration we have "initisalisable", "configurable" and "validatable"
- The `DiscordBot` struct is actually used properly now and in `main.go` it creates one and passes it the config file it found using `Configurable.Load`. Then it calls `DiscordBot.Validate` to make sure everything is setup correctly, all of the other structs that the `DiscordBot` stores must also implement the `Validatable` interface because it calls `Validate` on each of those as well. If everything is happy then at this point everything is ready to go, now we call `Initialise` from the `Initialisable` interface which creates the database connection and connects to the discord API.
Other changes:
- I made a `BotDatabase` interface which is implemented by our `MongoDB` struct, if we wanted to support other database methods then that would be easy to add

There's a bunch of other stuff I changed but I can't be arsed to write it now.

TL:DR, everything is now configurable from a JSON file

### CODE REVIEW

**Comment by ashketchupppp** [comment]
Wow! I'm blown away by this excellent code!
